app/lib/actions/candlepin/environment/set_content.rb

```diff
@@ -19,20 +19,46 @@ class SetContent < Candlepin::Abstract
           param :content_ids, Array
         end
 
-        def run
-          saved_cp_ids = ::Katello::Resources::Candlepin::Environment.
+        def existing_ids
+          ::Katello::Resources::Candlepin::Environment.
               find(input[:cp_environment_id])[:environmentContent].map do |content|
             content[:contentId]
           end
+        end
+
+        def run
+          saved_cp_ids = existing_ids
           add_ids    = input[:content_ids] - saved_cp_ids
           delete_ids = saved_cp_ids - input[:content_ids]
-
-          output[:add_ids]         =    add_ids
-          unless add_ids.empty?
-            output[:add_response]    = ::Katello::Resources::Candlepin::Environment.
-              add_content(input[:cp_environment_id], add_ids)
+          retries = 0
+          max_retries = 2
+          completed = false
+          until completed
+            output[:add_ids]         =    add_ids
+            unless add_ids.empty?
+              begin
+                output[:add_response]    = ::Katello::Resources::Candlepin::Environment.
+                  add_content(input[:cp_environment_id], add_ids)
+                completed = true
+              rescue RestClient::InternalServerError => e
+                # HACK
+                # Candlepin error messages do not raise any specific error for duplicate content ids
+                # the error raised looks something like this
+                # RestClient::InternalServerError: Katello::Resources::Candlepin::Environment: 500 Internal Server Error
+                # {"displayMessage":"Runtime Error could not execute statement at
+                # org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse:2,102",
+                # "requestUuid":"13c15fcd-0bc2-4686-a988-e0ba70377b86"} (POST /candlepin/environments/1/content)
+                # so as pure conjecture we are using this and just guessing if its a dup id.
+                if e.message.include?("org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse")
+                  add_ids = input[:content_ids] - existing_ids
```

Comment:
Code has allowance for that too . so see the line that says
{{{
                if e.message.include?("org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse")
                  add_ids = input[:content_ids] - existing_ids
                else
}}}
that existing_ids does a fresh get. Hmm sounds like your idea might work. Actually I am only catching 500s. 404 wlll raise a different exception anyway.